packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -238,43 +285,142 @@ contract FundingRateStore is FundingRateStoreInterface, Testable, Lockable {
                 fundingRateDispute.proposal.proposer,
-                fundingRateDispute.proposal.finalFee.rawValue
+                proposerReward.add(proposerRebate).rawValue
             );
         }
 
         // Update current rate to settlement rate if there has not been a published funding rate since the dispute
         // began.
         FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(perpetual);
-        if (_getLatestProposeTimeForContract(perpetual) <= proposalTime) {
+        if (fundingRateRecord.proposeTime <= proposalTime) {
             fundingRateRecord.rate = settlementRate;
             fundingRateRecord.proposeTime = proposalTime;
-            emit PublishedRate(perpetual, settlementRate.rawValue, proposalTime, proposer);
+            // Note: Set rewards to 0 since we published the DVM resolved funding rate.
+            emit PublishedRate(perpetual, settlementRate.rawValue, proposalTime, proposer, 0, 0);
         }
 
         // Delete dispute
         delete fundingRateDisputes[perpetual][proposalTime];
     }
 
+    /**
+     * @notice Publishes any expired pending proposals and pays out proposal rewards if neccessary.
+     * @param perpetual Contract to check proposed funding rates for.
+     */
+    function withdrawProposalRewards(address perpetual) external nonReentrant() publishAndWithdrawProposal(perpetual) {}
+
+    /**
+     * @notice Set the reward rate (per second) for a specific `perpetual` contract.
+     * @dev Callable only by the Perpetual contract.
+     */
+    function setRewardRate(address perpetual, FixedPoint.Unsigned memory rewardRate)
+        external
+        override
+        nonReentrant()
+        publishAndWithdrawProposal(perpetual)
+    {
+        require(msg.sender == perpetual, "Caller not perpetual");
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(perpetual);
+        fundingRateRecord.rewardRatePerSecond = rewardRate;
+
+        // Set last propose time to current time since rewards will start accruing from here on out.
+        fundingRateRecord.proposeTime = getCurrentTime();
+
+        emit ChangedRewardRate(perpetual, rewardRate.rawValue);
+    }
+
+    /**
+     * @notice Helpful method for proposers who want to calculate their potential rewards and useful for testing.
+     */
+    function calculateProposalRewardPct(
+        address perpetual,
+        uint256 startTime,
+        uint256 endTime,
+        FixedPoint.Signed memory proposedRate,
+        FixedPoint.Signed memory currentRate
+    ) external view returns (FixedPoint.Unsigned memory reward) {
+        return _calculateProposalRewardPct(perpetual, startTime, endTime, proposedRate, currentRate);
+    }
+
     /****************************************
      *         INTERNAL FUNCTIONS           *
      ****************************************/
 
+    // TODO: Do we need to add a reentrancy guard to this method because it makes external calls?
```

Comment:
yes. I think it should have the guard. could leave this for a separate PR where we go through again double checking all reentrancy points.

Reply:
I added the guard. In other contracts we don't add this guard to non-external methods so we should think more about this in another PR

Reply:
I don't think internal methods should have guards since they are only called by external methods that _do_ have guards. So they are effectively guarded. If you want to emphasize that this method is protected from re-entrancy, it might be worth putting a comment on the method that callers are expected to be guarded.

Reply:
Good point, I'm not gonna add a comment however because there are a lot of internal methods in the Perp system that should also have that same warning, so let's just say its implied for now

Reply:
SGTM!